CTFd/__init__.py

```diff
@@ -100,35 +100,31 @@ def _load_template(self, name, globals):
 
 
 class ThemeLoader(FileSystemLoader):
-    """Custom FileSystemLoader that switches themes based on the configuration value"""
-
-    def __init__(self, searchpath, encoding="utf-8", followlinks=False):
+    """Custom FileSystemLoader that is aware of theme structure and config.
+    """
+
+    DEFAULT_THEMES_PATH = os.path.join(os.path.dirname(__file__), "themes")
+    _ADMIN_THEME_PREFIX = ADMIN_THEME + "/"
+
+    def __init__(
+        self,
+        searchpath=DEFAULT_THEMES_PATH,
+        theme_name=None,
+        encoding="utf-8",
+        followlinks=False,
+    ):
         super(ThemeLoader, self).__init__(searchpath, encoding, followlinks)
-        self.overriden_templates = {}
+        self.theme_name = theme_name
 
     def get_source(self, environment, template):
-        # Check if the template has been overriden
-        if template in self.overriden_templates:
-            return self.overriden_templates[template], template, lambda: True
-
-        # Check if the template requested is for the admin panel
-        if template.startswith("admin/"):
-            template = template[6:]  # Strip out admin/
-            template = "/".join(["admin", "templates", template])
-            return super(ThemeLoader, self).get_source(environment, template)
-
-        # Load regular theme data with potential fallbacks
-        first_cand, *others = ctf_theme_candidates()
-        for cand_theme in (first_cand, *others):
-            tpl_path = safe_join(cand_theme, "templates", template)
-            try:
-                return super(ThemeLoader, self).get_source(environment, tpl_path)
-            except jinja2.exceptions.TemplateNotFound:
-                pass
-        else:
-            # If we found no matching templates then raise an exception for the
-            # first one we intended to try
-            raise jinja2.exceptions.TemplateNotFound(first_cand)
+        # Refuse to load `admin/*` from a loader not for the admin theme
+        if template.startswith(self._ADMIN_THEME_PREFIX):
+            if self.theme_name != ADMIN_THEME:
+                raise jinja2.TemplateNotFound(template)
+            template = template[len(self._ADMIN_THEME_PREFIX) :]
+        theme_name = self.theme_name or str(utils.get_config("ctf_theme"))
```

Comment:
Can this not be set in the constructor if `theme_name` is `None`?

Reply:
It looks like you must have realised why this is done each time, but just to answer the question, it's to avoid ignoring a change of theme at runtime :) Being able to alter the `theme_name` of the default `ThemeLoader` instance (ie. the one with no `theme_name` set) when we alter the config would be useful but IIUC this is done by some indirect logic so working out when the `ctf_theme` option changes specifically might be annoying to do. If we did manage to do that though, the change you suggested on L117 would be exactly what we'd want to do.